### PR TITLE
2363/3

### PR DESCRIPTION
**Pull request by dixent**
## Номер
2363

## Номер задания
3

## Ссылка на видео с демо
https://youtu.be/RL9IJm2VOok

## Комментарии
Сделал, надеюсь неплохо

### CODE REVIEW

File: 2363/3/config.ru [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Bundler.require(:default, :development)
+Dir.glob('./{controllers,helpers,models}/*.rb').sort.each { |file| require file }
+Post.redis = Redic.new('redis://127.0.0.1:6379/0')
+Comment.redis = Redic.new('redis://127.0.0.1:6379/1')
+
+use Rack::MethodOverride
```
Comment (dzhlobo) [inline 2363/3/config.ru]: Why do you need this middleware?

File: 2363/3/views/header.slim [hunk 2]
```diff
@@ -0,0 +1,5 @@
+  form action='post/delete_all' method='post'
+    div class="btn-group btn-group-toggle" data-toggle="buttons"
+      a role='button' class='btn btn-outline-dark' href='/' Main
+      a role='button' class='btn btn-outline-dark' href='add' Add link
+      button value='delete' name='_method' id='hidden' type='submit' class='btn btn-outline-dark' onclick="return confirm('Delete all?')" Delete all
```
Comment (dzhlobo) [inline 2363/3/views/header.slim]: Please setup [editorconfig](https://editorconfig.org/) for your editor and add a `.editorconfig` file to your folder so your editor will always add newline at the end of file.

File: 2363/3/sinatra.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+gem 'shotgun'
+gem 'thin'
+--servers=thin
+padrino sintra models 
+gem sitekick мультипоточность
```
Comment (dzhlobo) [inline 2363/3/sinatra.txt]: Huh, why do you need this file? Isn't it better move this to README?

File: 2363/3/controllers/posts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    @post = Post.all[params[:id]]
+    @post.delete
+    redirect '/'
+  end
+
+  delete '/post/delete_all' do
+    Post.all.each(&:delete)
+    redirect '/'
+  end
+end
```
Comment (dzhlobo) [inline 2363/3/controllers/posts_controller.rb]: Let's use some kind of convention for routes just to make our code a bit cleaner:

* `GET /posts` -- show all posts
* `GET /posts/new` -- show a page to add a new post
* `POST /posts` -- creates a new post
* `GET /posts/:id` -- shows one post
* `DELETE /posts/:id` -- deletes one post
* `DELETE /posts` -- deletes all posts

It's very common convention and the one you will use in rails apps: https://guides.rubyonrails.org/routing.html#crud-verbs-and-actions.

File: 2363/3/config.ru (hunk 1 above)
Comment (dixent) [reply 2363/3/config.ru]: For delete request

File: 2363/3/sinatra.txt (hunk 3 above)
Comment (dixent) [reply 2363/3/sinatra.txt]: Ops, that txt from lecture about sinatra

File: 2363/3/Gemfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+gem 'pry', group: :development
+
+group :development, :production do
+  gem 'config'
+  gem 'json'
+  gem 'mechanize'
+  gem 'ohm'
+  gem 'rake'
+  gem 'shotgun'
```
Comment (allomov) [inline 2363/3/Gemfile]: shotgun also goes to development group

File: 2363/3/Gemfile [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  gem 'sinatra'
+  gem 'sinatra-config-file', '~> 1.0'
```
Comment (allomov) [inline 2363/3/Gemfile]: perfect

File: 2363/3/controllers/posts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    Post.all.each do |post|
+      post.delete if post.link == params[:link]
+    end
+    comments = CommentsParser.new(params[:link]).run
+    rating = CountRating.new(comments, settings.access_key).run
+    post = Post.create link: params[:link], rating: rating.sum / rating.size
+    comments.zip(rating).each do |obj|
+      post.comments.add(Comment.create(text: obj.first, rating: obj.last))
+    end
+    redirect '/posts'
```
Comment (allomov) [inline 2363/3/controllers/posts_controller.rb]: the method is too large, but for now it's fine

File: 2363/3/controllers/posts_controller.rb (hunk 4 above)
Comment (allomov) [reply 2363/3/controllers/posts_controller.rb]: у них еще не было лекции про REST, так что я разрешаю придумывать свой интерфейс. но знание REST это хорошо и важно.

**Pull request merged**